Add context menu's

UnderKoen opened the pull request:
<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes -->

Currently only on budget page and transactions

Still todo
- [x] Schedules
- [x] Payees
- [x] Rules

Teprifer commented:
Had a sneak peak, Chrome on w11:

Transaction table: the right click menu isn't dismissed on left click, so it's possible to right click and bring up a menu, left click to try dismiss(doesn't), then right click again and get two multiple menus:

![image](https://github.com/user-attachments/assets/4034418d-242e-4201-8037-2329d4996ac0)
Can also reproduce if the second click is a right click outside the transaction table.

Budget table:
Right clicking on one category prevents right clicking on another until the menu is dismissed. Although left click does dismiss it here.

jfdoming commented:
This is really cool! Quick observation: I noticed that the positioning of the menu is not quite under the cursor—for transactions, rules, and payees, it's a few pixels below it, and for schedules, it seems to always be at the bottom of the entry. I think folks are more likely used to context menus appearing directly at the cursor. Would it be possible to make it behave that way instead?

UnderKoen commented:
@jfdoming I changed the positioning on the context menu. I didn't do this on the budget page because it didn't really work.

youngcw commented:
Is it worth having context menus in the budget table?  They were already viewable with a regular click.

I think it would be worthwile to have the budget table bulk edit options available in the context menu when multiple transactions are selected

UnderKoen commented:
> Is it worth having context menus in the budget table?

My intuition is to right click to see the option. I constantly right click when wanting to do these actions.

UnderKoen commented on packages/desktop-client/src/components/ManageRules.tsx:
```diff
@@ -203,6 +203,13 @@ function ManageRulesContent({
     setLoading(false);
   }
 
+  async function onDeleteRule(id: string) {
+    setLoading(true);
+    await send('rule-delete', id);
+    await loadRules();
+    setLoading(false);
+  }
```
The error that could happen is that it is possible to have a schedule rule selected and to try and delete that rule. As the context menu doesn't show this option this is not a problem.

UnderKoen commented on packages/desktop-client/src/components/budget/envelope/EnvelopeBudgetComponents.tsx:
```diff
@@ -266,6 +271,7 @@ export const ExpenseCategoryMonth = memo(function ExpenseCategoryMonth({
               isOpen={budgetMenuOpen}
               onOpenChange={() => setBudgetMenuOpen(false)}
               style={{ width: 200 }}
+              isNonModal
```
I don't think this should be included in this PR as it would change much.

UnderKoen commented on packages/desktop-client/src/components/transactions/TransactionMenu.tsx:
```diff
@@ -0,0 +1,139 @@
+  const canUnsplitTransactions =
+    !transaction.reconciled && (transaction.is_parent || transaction.is_child);
+
+  function onViewSchedule() {
+    const firstId = transaction.id;
+    let scheduleId;
+    if (isPreviewId(firstId)) {
+      const parts = firstId.split('/');
+      scheduleId = parts[1];
+    } else {
+      scheduleId = transaction.schedule;
+    }
+
+    if (scheduleId) {
+      dispatch(pushModal('schedule-edit', { id: scheduleId }));
+    }
+  }
```
When `isPreviewId` is true `firstId` always has a `/`

UnderKoen commented on packages/desktop-client/src/components/payees/PayeeTableRow.tsx:
```diff
@@ -122,7 +135,58 @@ export const PayeeTableRow = memo(
+        >
+          <Menu
+            items={[
+              payee.transfer_acct == null && {
+                name: 'delete',
```
When checking with `null` I prefer using `==` as it also matches `undefined`. Which is in this project plausable because TS checking is not as strict.

UnderKoen commented on packages/desktop-client/src/components/payees/PayeeTableRow.tsx:
```diff
@@ -122,7 +135,58 @@
+                  onDelete(id);
+                  break;
+                case 'favorite':
+                  onUpdate(id, 'favorite', payee.favorite ? 0 : 1);
```
The function expects 1 or 0

UnderKoen commented on packages/desktop-client/src/components/payees/PayeeTableRow.tsx:
```diff
@@ -122,7 +135,58 @@
+                  onCreateRule(id);
+                  break;
+                default:
+                  throw new Error(`Unrecognized menu option: ${name}`);
+              }
```
I have copy this from other implementations, so I assume this is the standart.

UnderKoen commented on packages/desktop-client/src/components/transactions/TransactionMenu.tsx:
```diff
@@ -0,0 +1,139 @@
+  return (
+    <Menu
+      {...props}
+      onMenuSelect={name => {
+        switch (name) {
+          case 'duplicate':
+            onDuplicate(transaction.id);
+            break;
+          case 'delete':
+            onDelete(transaction.id);
+            break;
+          case 'unsplit-transactions':
+            onMakeAsNonSplitTransactions(transaction.id);
+            break;
+          case 'post-transaction':
+          case 'skip':
+            onScheduleAction(name, transaction.id);
+            break;
+          case 'view-schedule':
+            onViewSchedule();
+            break;
+          case 'link-schedule':
+            onLinkSchedule(transaction.id);
+            break;
+          case 'unlink-schedule':
+            onUnlinkSchedule(transaction.id);
+            break;
+          case 'create-rule':
+            onCreateRule(transaction.id);
+            break;
+          default:
+            throw new Error(`Unrecognized menu option: ${name}`);
+        }
+        closeMenu();
+      }}
```
I have copy this from other implementations, so I assume this is the standart.

UnderKoen commented on packages/desktop-client/src/components/payees/PayeeTableRow.tsx:
```diff
@@ -122,7 +135,58 @@
+        >
+          <Menu
+            items={[
+              payee.transfer_acct == null && {
+                name: 'delete',
+                text: t('Delete'),
+              },
+              payee.transfer_acct == null && {
+                name: 'favorite',
+                text: payee.favorite ? t('Unfavorite') : t('Favorite'),
+              },
+              ruleCount > 0 && { name: 'view-rules', text: t('View rules') },
+              { name: 'create-rule', text: t('Create rule') },
+            ]}
```
This doesn't matter here.

UnderKoen commented on packages/desktop-client/src/components/payees/PayeeTableRow.tsx:
```diff
@@ -122,7 +135,58 @@
+          setMenuOpen(true);
+          const rect = e.currentTarget.getBoundingClientRect();
+          setCrossOffset(e.clientX - rect.left);
+          setOffset(e.clientY - rect.bottom);
```
Due to the alignment of the menu this is correct.

UnderKoen commented on packages/desktop-client/src/components/rules/RuleRow.tsx:
```diff
@@ -59,7 +80,45 @@ export const RuleRow = memo(
         collapsed={true}
         onMouseEnter={() => onHover && onHover(rule.id)}
         onMouseLeave={() => onHover && onHover(null)}
+        onContextMenu={e => {
+          e.preventDefault();
+          setMenuOpen(true);
+          const rect = triggerRef.current.getBoundingClientRect();
+          setCrossOffset(e.clientX - rect.left);
+          setOffset(e.clientY - rect.bottom);
+        }}
```
Due to the alignment of the menu this is correct.

UnderKoen commented on packages/desktop-client/src/components/schedules/SchedulesTable.tsx:
```diff
@@ -196,6 +169,125 @@ export function ScheduleAmountCell({
+      onContextMenu={e => {
+        if (minimal) return;
+        e.preventDefault();
+        const rect = e.currentTarget.getBoundingClientRect();
+        setCrossOffset(e.clientX - rect.left);
+        setOffset(e.clientY - rect.bottom);
+        setOpen('contextMenu');
```
Due to the alignment of the menu this is correct.

joel-jeremy commented on packages/desktop-client/src/components/accounts/Account.tsx:
```diff
@@ -1780,6 +1780,15 @@ class AccountInternal extends PureComponent<
                   sortField={this.state.sort?.field}
                   ascDesc={this.state.sort?.ascDesc}
                   onChange={this.onTransactionsChange}
+                  onBatchDelete={this.onBatchDelete}
```
There is a `useTransactionBatchActions` hook. Maybe we can use that here?

UnderKoen replied:
I just tried it. But unfortunatly the table doesn't update quite right. When you duplicate an transaction you need te reload the page to see the change.

jfdoming commented:
This is looking pretty good now! I found one issue on the budget table, related to the positioning of the menu:

https://github.com/user-attachments/assets/6165aba0-0a14-4a34-a1c0-ee3af3cb1cfb

Other than that, I think I'm happy to merge this and follow up with any fixes we notice.

UnderKoen commented:
Thanks @jfdoming I hadn't noticed it. I fixed it.

I also changed the positioning it felt more natural for me. 
![image](https://github.com/user-attachments/assets/82dd9d4c-5f00-47c7-92d4-9a27069f2d26)
Let me know what you think!

UnderKoen commented on packages/desktop-client/src/components/budget/SidebarGroup.tsx:
```diff
@@ -64,10 +64,16 @@ export function SidebarGroup({
         WebkitUserSelect: 'none',
+        height: 20,
       }}
+      ref={triggerRef}
       onClick={() => {
         onToggleCollapse(group.id);
       }}
+      onContextMenu={e => {
+        e.preventDefault();
+        setMenuOpen(true);
+      }}
```
This is done in another file

jfdoming reviewed (commented):
It's definitely looking better! I'd still prefer if we made it positioned at the mouse even on the budget page, but that can be for a future PR.

However, I think some things here may need more thought; in particular:
- Choosing an action via right-click when multiple items are selected in the budget table doesn't behave as expected (only applies to the transaction that was right-clicked)
- Some options are missing from the right-click menu on the transactions table (Make transfer) and the shortcuts are not displayed
- Interactions still occur when a right click menu opens (in a native menu, all interactions are normally disabled)

In the interest of not having to fix all of these at once, how do folks feel about putting this behind a feature flag? It's already pretty usable and useful, but I think we might get some questions about the behaviours I mentioned above.

Teprifer commented:
@jfdoming I agree with your assessment and on board with your feature flag suggestion. It'll also allow chipping away at those points in smaller pieces of work individually.

youngcw commented:
I don't know what happened, but none of the context menus are showing for me anymore.  Im testing on Brave, linux

UnderKoen commented:
I added a feature flag as requested by jfdoming

youngcw commented:
> I added a feature flag as requested by jfdoming

That was it.  I need to read better

youngcw reviewed (approved):
Since this is now behind a feature flag, I think we can push the other requested additions to a different PR.
@jfdoming @joel-jeremy Do you have any other comments?

jfdoming commented on packages/desktop-client/src/components/settings/Experimental.tsx:
```diff
@@ -122,6 +122,12 @@ export function ExperimentalFeatures() {
+              flag="contextMenus"
+              feedbackLink="https://github.com/actualbudget/actual/issues/3706"
+            >
+              <Trans>Context menu’s</Trans>
```
nit:
```suggestion
              <Trans>Context menus</Trans>
```

jfdoming reviewed (dismissed):
Looks good to me now that things are behind a feature flag. Noticed one non-blocking typo. Also another observation for a future PR: we should make the text within the context menu items unselectable.

UnderKoen commented:
**Fixxed after approval:**
- The button to budget wasn't opening secondary menu's. Now it does

cordlord commented:
Sorry if I missed it or this isn't the right place, but has anyone had success using this on Firefox?

All I get is the default context menu.

UnderKoen commented:
@cordlord this is currently behind a feature flag, you can enable this in the settings under experimental

cordlord commented:
@UnderKoen Bingo. Didn't even think to check here first. My bad. Thanks!
